easybuild/easyblocks/t/tensorflow.py

```diff
@@ -835,41 +838,98 @@ def test_step(self):
+            test_tag_filters_name = 'test_tag_filters_' + device
+            test_tag_filters = self.cfg[test_tag_filters_name]
+            if not test_tag_filters:
+                self.log.info('Skipping %s test because %s is not set', device, test_tag_filters_name)
+                continue
+            else:
+                self.log.info('Starting %s test', device)
+
+            current_test_opts = test_opts[:]
+            current_test_opts.append('--local_test_jobs=%s' % num_test_jobs[device])
+
+            # Add both build and test tag filters as done by the TF CI scripts
+            current_test_opts.extend("--%s_tag_filters='%s'" % (step, test_tag_filters) for step in ('test', 'build'))
+
+            # TF can only run tests explicitely marked as 'gpu' on GPUs and those don't work on cpu
+            # See https://github.com/tensorflow/tensorflow/issues/45664
+            if device == 'cpu':
+                current_test_opts.append('--test_env=CUDA_VISIBLE_DEVICES=-1')
+
+            # Append user specified options last
+            current_test_opts.append(cfg_testopts[device])
+
+            # Compose final command
+            cmd = ' '.join(
+                [self.cfg['pretestopts']]
+                + ['bazel']
+                + self.bazel_opts
+                + ['test']
+                + current_test_opts
+                + ['--']
+                # specify targets to test as last argument
+                + test_targets
+            )
+
+            stdouterr, ec = run_cmd(cmd, log_ok=False, simple=False)
+            if ec:
+                fail_msg = 'Tests on %s (cmd: %s) failed with exit code %s and output:\n%s' % (
+                    device, cmd, ec, stdouterr)
+                self.log.warning(fail_msg)
+                # Try to enhance error message
+                failed_tests = []
+                failed_test_logs = dict()
+                # Bazel outputs failed tests like "//tensorflow/c:kernels_test   FAILED in[...]"
+                for m in re.finditer(r'^(//[a-zA-Z_/:]+) + FAILED', stdouterr, re.MULTILINE):
+                    test_name = m.group(1)
+                    failed_tests.append(test_name)
+                    # Logs are in a folder named after the test, e.g. tensorflow/c/kernels_test
+                    test_folder = test_name[2:].replace(':', '/')
+                    test_log_re = re.compile(r'.*\n(.*\n)?\s*(/.*/testlogs/%s/test.log)' % test_folder)
+                    log_match = test_log_re.match(stdouterr, m.end())
+                    if log_match:
+                        failed_test_logs[test_name] = log_match.group(2)
+                # Print info about failed tests to log
+                for test_name, log_path in failed_test_logs.items():
+                    if os.path.exists(log_path):
+                        self.log.warning('Test %s failed with output\n%s', test_name,
+                                         read_file(log_path, log_error=False))
```

Comment:
Still log warning for failed test if log is not found?

Reply:
This is only to enhance what is reported by the EasyBuildError 3 lines below. I'll clarify the comment

Reply:
I think it's worth logging a warning that says "`Test X failed, but no log file found`" under `else:`...

Reply:
I think that would be excessive. IIRC if compilation fails then all tests are marked as failure and hence a line of that for all 600 tests or so...
Are you sure?

Reply:
If compilation fails then the tests will never be run, since EasyBuild will stop early?

I think it's worth mentioning that no log files were found for failing tests? Or is that not abnormal that they're missing?

Reply:
No, the test step of Bazel will compile everything needed for the tests, so at least the tests. And that can fail. Hence I would not say it is abnormal that they are missing.